base/range.jl

```diff
@@ -805,57 +754,40 @@ convert{T1,T2}(::Type{StepRange{T1,T2}}, r::Range) =
+promote_rule{T1,T2,R1,R2,S1,S2}(::Type{StepRangeLen{T1,R1,S1}},::Type{StepRangeLen{T2,R2,S2}}) =
+    StepRangeLen{promote_type(T1,T2), promote_type(R1,R2), promote_type(S1,S2)}
+convert{T,R,S}(::Type{StepRangeLen{T,R,S}}, r::StepRangeLen{T,R,S}) = r
+convert{T,R,S}(::Type{StepRangeLen{T,R,S}}, r::StepRangeLen) =
+    StepRangeLen{T,R,S}(convert(R, r.ref), convert(S, r.step), length(r), r.offset)
+convert{T}(::Type{StepRangeLen{T}}, r::StepRangeLen) =
+    StepRangeLen(convert(T, r.ref), convert(T, r.step), length(r), r.offset)
+
+promote_rule{T,R,S,OR<:Range}(::Type{StepRangeLen{T,R,S}}, ::Type{OR}) =
+    StepRangeLen{promote_type(T,eltype(OR)),promote_type(R,eltype(OR)),promote_type(S,eltype(OR))}
+convert{T,R,S}(::Type{StepRangeLen{T,R,S}}, r::Range) =
+    StepRangeLen{T,R,S}(R(first(r)), S(step(r)), length(r))
+convert{T}(::Type{StepRangeLen{T}}, r::Range) =
+    StepRangeLen(T(first(r)), T(step(r)), length(r))
+convert(::Type{StepRangeLen}, r::Range) = convert(StepRangeLen{eltype(r)}, r)
 
 promote_rule{T1,T2}(::Type{LinSpace{T1}},::Type{LinSpace{T2}}) =
     LinSpace{promote_type(T1,T2)}
-convert{T<:AbstractFloat}(::Type{LinSpace{T}}, r::LinSpace{T}) = r
-convert{T<:AbstractFloat}(::Type{LinSpace{T}}, r::LinSpace) =
-    LinSpace{T}(r.start, r.stop, r.len, r.divisor)
-
-promote_rule{F,OR<:OrdinalRange}(::Type{LinSpace{F}}, ::Type{OR}) =
-    LinSpace{promote_type(F,eltype(OR))}
-convert{T<:AbstractFloat}(::Type{LinSpace{T}}, r::OrdinalRange) =
-    linspace(convert(T, first(r)), convert(T, last(r)), convert(T, length(r)))
+convert{T}(::Type{LinSpace{T}}, r::LinSpace{T}) = r
+convert{T}(::Type{LinSpace{T}}, r::Range) =
+    LinSpace{T}(first(r), last(r), r.len)
+
+promote_rule{T,OR<:OrdinalRange}(::Type{LinSpace{T}}, ::Type{OR}) =
+    LinSpace{promote_type(T,eltype(OR))}
+convert{T}(::Type{LinSpace{T}}, r::OrdinalRange) =
+    linspace(convert(T, first(r)), convert(T, last(r)), length(r))
 convert{T}(::Type{LinSpace}, r::OrdinalRange{T}) =
     convert(LinSpace{typeof(float(first(r)))}, r)
 
-# Promote FloatRange to LinSpace
-promote_rule{F,OR<:FloatRange}(::Type{LinSpace{F}}, ::Type{OR}) =
-    LinSpace{promote_type(F,eltype(OR))}
-convert{T<:AbstractFloat}(::Type{LinSpace{T}}, r::FloatRange) =
-    linspace(convert(T, first(r)), convert(T, last(r)), convert(T, length(r)))
-convert{T<:AbstractFloat}(::Type{LinSpace}, r::FloatRange{T}) =
-    convert(LinSpace{T}, r)
-
+promote_rule{L,T,R,S}(::Type{LinSpace{L}}, ::Type{StepRangeLen{T,R,S}}) =
+    StepRangeLen{promote_type(L,T),promote_type(L,R),promote_type(L,S)}
 
 # +/- of ranges is defined in operators.jl (to be able to use @eval etc.)
 
-## non-linear operations on ranges and fallbacks for non-real numbers ##
-
-+(x::Number, r::Range) = [ x+y for y=r ]
-+(r::Range, y::Number) = [ x+y for x=r ]
-
--(x::Number, r::Range) = [ x-y for y=r ]
--(r::Range, y::Number) = [ x-y for x=r ]
-
-*(x::Number, r::Range) = [ x*y for y=r ]
```

Comment:
should this abstract fallback be deprecated? you only have number times concrete range types above